File: Backend-JS/reel-service/src/models/VideoTutorial.js
const mongoose = require("mongoose");
const { VIDEO_LIMITS, VIDEO_TYPES, VISIBILITY } = require('../utils/constants');

const VideoTutorialSchema = new mongoose.Schema(
  {
    userId: {
      type: String,
      required: true,
      index: true,
    },
    userName: {
      type: String,
      required: true,
    },
    profilePhoto: {
      type: String,
    },
    title: {
      type: String,
      required: true,
      trim: true,
      maxLength: VIDEO_LIMITS.MAX_TITLE_LENGTH,
    },
    description: {
      type: String,
      trim: true,
      maxLength: VIDEO_LIMITS.MAX_DESCRIPTION_LENGTH,
    },
    thumbnail: {
      type: String,
      required: true,
    },
    videoUrl: {
      type: String,
      required: true,
    },
    videoType: {
      type: String,
      enum: Object.values(VIDEO_TYPES),
      required: true,
    },
    duration: {
      type: Number,
    },
    tags: {
      type: [String],
      validate: [arr => arr.length <= VIDEO_LIMITS.MAX_TAGS, `Maximum ${VIDEO_LIMITS.MAX_TAGS} tags allowed`],
    },
    category: {
      type: String,
      required: true,
      index: true,
    },
    visibility: {
      type: String,
      enum: Object.values(VISIBILITY),
      default: VISIBILITY.PUBLIC,
      index: true,
    },
    likes: {
      count: {
        type: Number,
        default: 0,
        index: true,
      },
      users: [{
        type: String, // userId
      }],
    },
    views: {
      count: {
        type: Number,
        default: 0,
        index: true,
      },
      unique: [{
        type: String, // userId - limit to last 1000 for performance
      }],
    },
    comments: {
      count: {
        type: Number,
        default: 0,
      },
    },
    reports: {
      type: [{
        userId: String,
        reason: String,
        description: String,
        date: {
          type: Date,
          default: Date.now,
        },
      }],
      default: [],
    },
    isActive: {
      type: Boolean,
      default: true,
      index: true,
    },
  },
  {
    timestamps: true,
  }
);

// Compound indexes for common query patterns (optimized for 10k users)
VideoTutorialSchema.index({ userId: 1, createdAt: -1 }); // User's videos
VideoTutorialSchema.index({ title: 'text', description: 'text', tags: 'text' }); // Text search
VideoTutorialSchema.index({ 'likes.count': -1, createdAt: -1 }); // Popular videos
VideoTutorialSchema.index({ 'views.count': -1, createdAt: -1 }); // Most viewed
VideoTutorialSchema.index({ category: 1, createdAt: -1 }); // Videos by category
VideoTutorialSchema.index({ visibility: 1, isActive: 1, createdAt: -1 }); // Public feed
VideoTutorialSchema.index({ category: 1, 'views.count': -1 }); // Popular by category
VideoTutorialSchema.index({ tags: 1 }); // Videos by tag

// Pre-save hook to limit unique views array
VideoTutorialSchema.pre('save', function(next) {
  // Keep only last 1000 unique viewers to prevent unbounded growth
  if (this.views.unique && this.views.unique.length > 1000) {
    this.views.unique = this.views.unique.slice(-1000);
  }
  // Check reports threshold
  if (this.reports && this.reports.length >= VIDEO_LIMITS.MAX_REPORTS_BEFORE_REVIEW) {
    // Flag for review
    this.isActive = false;
  }
  next();
});

// Static method to get videos by category
VideoTutorialSchema.statics.getByCategory = function(category, { page = 1, limit = 12, sort = 'recent' } = {}) {
  const skip = (page - 1) * limit;
  const sortOptions = {
    recent: { createdAt: -1 },
    popular: { 'views.count': -1 },
    trending: { 'likes.count': -1 },
  };

  return this.find({ category, visibility: VISIBILITY.PUBLIC, isActive: true })
    .sort(sortOptions[sort] || sortOptions.recent)
    .skip(skip)
    .limit(limit)
    .lean();
};

// Static method to search videos
VideoTutorialSchema.statics.searchVideos = function(query, { page = 1, limit = 12 } = {}) {
  const skip = (page - 1) * limit;
  return this.find({
    $text: { $search: query },
    visibility: VISIBILITY.PUBLIC,
    isActive: true,
  })
    .sort({ score: { $meta: 'textScore' } })
    .skip(skip)
    .limit(limit)
    .lean();
};

// Static method to get related videos
VideoTutorialSchema.statics.getRelated = function(videoId, category, tags, limit = 8) {
  return this.find({
    _id: { $ne: videoId },
    isActive: true,
    visibility: VISIBILITY.PUBLIC,
    $or: [
      { category },
      { tags: { $in: tags } },
    ],
  })
    .sort({ 'views.count': -1 })
    .limit(limit)
    .lean();
};

// Instance method to record view
VideoTutorialSchema.methods.recordView = async function(userId) {
  if (userId && !this.views.unique.includes(userId)) {
    this.views.count += 1;
    this.views.unique.push(userId);
    await this.save();
    return true;
  }
  return false;
};

module.exports = mongoose.model("VideoTutorial", VideoTutorialSchema);

File: Frontend/krishimantra/lib/data/models/video_tutorial.dart
class VideoTutorial {
  final String id;
  final String userId;
  final String userName;
  final String? profilePhoto;
  final String title;
  final String? description;
  final String thumbnail;
  final String videoUrl;
  final String videoType;
  final int? duration;
  final List<String> tags;
  final String category;
  final String visibility;
  final VideoStats likes;
  final VideoStats views;
  final VideoStats comments;
  final DateTime createdAt;
  final DateTime updatedAt;

  VideoTutorial({
    required this.id,
    required this.userId,
    required this.userName,
    this.profilePhoto,
    required this.title,
    this.description,
    required this.thumbnail,
    required this.videoUrl,
    required this.videoType,
    this.duration,
    required this.tags,
    required this.category,
    required this.visibility,
    required this.likes,
    required this.views,
    required this.comments,
    required this.createdAt,
    required this.updatedAt,
  });

  factory VideoTutorial.fromJson(Map<String, dynamic> json) {
    // Handle potential null or missing values in the JSON
    final likes =
        json['likes'] is Map ? json['likes'] : {'count': 0, 'users': []};
    final views =
        json['views'] is Map ? json['views'] : {'count': 0, 'unique': []};
    final comments = json['comments'] is Map ? json['comments'] : {'count': 0};

    return VideoTutorial(
      id: json['_id'] ?? '',
      userId: json['userId'] ?? '',
      userName: json['userName'] ?? 'Unknown',
      profilePhoto: json['profilePhoto'],
      title: json['title'] ?? 'Untitled',
      description: json['description'],
      thumbnail: json['thumbnail'] ?? '',
      videoUrl: json['videoUrl'] ?? '',
      videoType: json['videoType'] ?? 'unknown',
      duration: json['duration'],
      tags: List<String>.from(json['tags'] ?? []),
      category: json['category'] ?? 'Uncategorized',
      visibility: json['visibility'] ?? 'public',
      likes: VideoStats.fromJson({
        'count': likes['count'] ?? 0,
        'users': likes['users'] ?? [],
      }),
      views: VideoStats.fromJson({
        'count': views['count'] ?? 0,
        'unique': views['unique'] ?? [],
      }),
      comments: VideoStats.fromJson({
        'count': comments['count'] ?? 0,
        'users': [],
      }),
      createdAt:
          DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String()),
      updatedAt:
          DateTime.parse(json['updatedAt'] ?? DateTime.now().toIso8601String()),
    );
  }
}

class VideoStats {
  final int count;
  final List<String> users;

  VideoStats({
    required this.count,
    required this.users,
  });

  factory VideoStats.fromJson(Map<String, dynamic> json) {
    // Handle count that might come as either int, double, or String
    final dynamic rawCount = json['count'] ?? 0;
    int count;

    if (rawCount is int) {
      count = rawCount;
    } else if (rawCount is double) {
      count = rawCount.toInt();
    } else if (rawCount is String) {
      count = int.tryParse(rawCount) ?? 0;
    } else {
      count = 0;
    }

    return VideoStats(
      count: count,
      users: List<String>.from(json['users'] ?? json['unique'] ?? []),
    );
  }
}

class Comment {
  final String id;
  final String videoId;
  final String userId;
  final String userName;
  final String? profilePhoto;
  final String content;
  final VideoStats likes;
  final String? parentComment;
  final List<Comment> replies;
  final int depth;
  final bool isDeleted;
  final DateTime createdAt;
  final DateTime updatedAt;

  Comment({
    required this.id,
    required this.videoId,
    required this.userId,
    required this.userName,
    this.profilePhoto,
    required this.content,
    required this.likes,
    this.parentComment,
    required this.replies,
    required this.depth,
    required this.isDeleted,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Comment.fromJson(Map<String, dynamic> json) {
    return Comment(
      id: json['_id'],
      videoId: json['videoId'],
      userId: json['userId'],
      userName: json['userName'],
      profilePhoto: json['profilePhoto'],
      content: json['content'],
      likes: VideoStats.fromJson(json['likes']),
      parentComment: json['parentComment'],
      replies: (json['replies'] as List?)
              ?.map((reply) => Comment.fromJson(reply))
              .toList() ??
          [],
      depth: json['depth'] ?? 0,
      isDeleted: json['isDeleted'] ?? false,
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt: DateTime.parse(json['updatedAt']),
    );
  }
}
